src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -223,73 +251,133 @@ public Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Cance
-            if (changedSolution == null)
-            {
-                return Array.Empty<CodeActionOperation>();
-            }
-
-            return new CodeActionOperation[] { new ApplyChangesOperation(changedSolution) };
+            var changedSolution = await GetChangedSolutionAsync(CodeAnalysisProgress.None, cancellationToken).ConfigureAwait(false);
+            return changedSolution == null
+                ? Array.Empty<CodeActionOperation>()
+                : SpecializedCollections.SingletonEnumerable<CodeActionOperation>(new ApplyChangesOperation(changedSolution));
         }
 
-        internal virtual async Task<ImmutableArray<CodeActionOperation>> ComputeOperationsAsync(
-            IProgressTracker progressTracker, CancellationToken cancellationToken)
+#pragma warning disable RS0030 // Do not use banned APIs
+
+        /// <summary>
+        /// Override this method if you want to implement a <see cref="CodeAction"/> subclass that includes custom <see
+        /// cref="CodeActionOperation"/>'s.  Prefer overriding this method over <see
+        /// cref="ComputeOperationsAsync(CancellationToken)"/> when computation is long running and progress should be
+        /// shown to the user.
+        /// </summary>
+        protected virtual async Task<ImmutableArray<CodeActionOperation>> ComputeOperationsAsync(
+            IProgress<CodeAnalysisProgress> progress, CancellationToken cancellationToken)
         {
-            var operations = await ComputeOperationsAsync(cancellationToken).ConfigureAwait(false);
-            return operations.ToImmutableArrayOrEmpty();
+            // If the subclass overrode `ComputeOperationsAsync(CancellationToken)` then we must call into that in
+            // order to preserve whatever logic our subclass had had for determining the new solution.
+            if (IsNonProgressComputeOperationsAsyncOverridden())
+            {
+                var operations = await ComputeOperationsAsync(cancellationToken).ConfigureAwait(false);
+                return operations.ToImmutableArrayOrEmpty();
+            }
+            else
+            {
+                var changedSolution = await GetChangedSolutionAsync(progress, cancellationToken).ConfigureAwait(false);
+                return changedSolution == null
+                    ? ImmutableArray<CodeActionOperation>.Empty
+                    : ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(changedSolution));
+            }
         }
 
+#pragma warning restore RS0030 // Do not use banned APIs
+
         /// <summary>
         /// Override this method if you want to implement a <see cref="CodeAction"/> that has a set of preview operations that are different
-        /// than the operations produced by <see cref="ComputeOperationsAsync(CancellationToken)"/>.
+        /// than the operations produced by <see cref="ComputeOperationsAsync(IProgress{CodeAnalysisProgress}, CancellationToken)"/>.
```

Comment:
>IProgress{CodeAnalysisProgress},  [](http://example.com/codeflow?start=78&length=33)

Intentionally not added a version of ComputePreviewOperationsAsync that takes in an IProgress?

Reply:
correct.  previews have no way to support progress (today at least).  It's *possibly* something we could consider in the future (and we'd do similar stuff here).  But it's not a big deal.